### PR TITLE
feat: add ring creation and signing

### PR DESCRIPTION
**Pull request by h5law**

### CODE REVIEW

File: relayer/proxy/proxy.go [hunk 1]
```diff
@@ -137,7 +139,25 @@ func (proxy *Proxy) signResponse(relayResponse *servicerTypes.RelayResponse) (si
...
+	// TODO: add ring comparison (https://github.com/noot/ring-go/issues/13)
+
+	sigBz, err := relayRequest.GetSignableBytes()
+	if err != nil {
+		return fmt.Errorf("failed to get request's signable bytes: %w", err)
+	}
+
+	ringSig := new(ring.RingSig)
+	crv := ring_secp256k1.NewCurve()
+	if err := ringSig.Deserialize(crv, relayRequest.Signature); err != nil {
```
Comment (red-0ne) [inline relayer/proxy/proxy.go:152]: Not sure what this is doing... Does this mean that `ringSig` is stateful?

**Review by red-0ne — COMMENTED** [review]
Great work Harry. exciting feature you landed here.

Left a few comments and wanted us to have more encapsulation regarding the signer. Application and Gateway signers are indeed different beasts but I think there is still room for encapsulation and abstraction.

Precisely, we should try to be able to not see the difference in their usage. The ring signer could carry the delegation client and handle ring reconstruction outside of the appclient. This will imply that we will construct them differently where we could have an internal `getSigner` method that in the case of an application; it would return the `ApplicationAddress` while on the ringsig variant it would extract it from the `RelayRequest` itself.

Let me know your thoughts on that.

File: x/application/keeper/msg_server_undelegate_from_portal.go [hunk 2]
```diff
@@ -17,6 +17,12 @@ func (k msgServer) UndelegateFromPortal(goCtx context.Context, msg *types.MsgUnd
 		logger.Error(fmt.Sprintf("could not update store with delegated portal for application: %s", msg.AppAddress))
 		return nil, err
 	}
+	if err := ctx.EventManager().EmitTypedEvent(&types.EventDelegate{
```
[the remark this answers is not in the document]
Comment (h5law) [reply x/application/keeper/msg_server_undelegate_from_portal.go:20]: The purpose of this event is to say a change of delegation has successfully occured to the address and we need to update our rings - maybe a name change on the event is needed. It doesn't matter if its delegation or undelegation as both require the ring to be fetched again.

File: relayer/proxy/proxy.go (hunk 1 above)
Comment (h5law) [reply relayer/proxy/proxy.go:152]: The `RingSig` struct contains the ring and public keys and this is needed for verification. This serialisation/deserialisation step enables us to transport the ring sig for verification.

```go
// RingSig represents a ring signature.
type RingSig struct {
	ring  *Ring          // array of public keys
	c     types.Scalar   // ring signature challenge
	s     []types.Scalar // ring signature values
	image types.Point    // key image
}
```

File: relayer/proxy/proxy.go [hunk 3]
```diff
@@ -137,7 +139,25 @@ func (proxy *Proxy) signResponse(relayResponse *servicerTypes.RelayResponse) (si
...
+	// TODO: cache ring for application at start of session
+	// TODO: add ring comparison (https://github.com/noot/ring-go/issues/13)
+
+	sigBz, err := relayRequest.GetSignableBytes()
+	if err != nil {
+		return fmt.Errorf("failed to get request's signable bytes: %w", err)
+	}
+
+	ringSig := new(ring.RingSig)
+	crv := ring_secp256k1.NewCurve()
```
[the remark this answers is not in the document]
Comment (h5law) [reply relayer/proxy/proxy.go:151]: AFAIK we will only support 1 key type, `secp256k1` however if this is not the case I can add a method that checks which key type it is and uses the corresponding curve. @Olshansk do you have more context here?

**Pull request merged**